### ISSUE TITLE
Export 'chai' into globals

### ISSUE THREAD

**Issue by andvgal** [trufflesuite/truffle#2090]
- [x] I've asked for help in the [Truffle Gitter](http://gitter.im/Consensys/truffle) before filing this issue.

---------------------------

## Issue

Related #1458 
A simple fix is to add one line at https://github.com/trufflesuite/truffle/blob/8013923d4a25585981d86858f4d979ac1f001fb4/packages/truffle-core/lib/test.js#L237

## Steps to Reproduce

Use `chai` directly in any test.

## Expected Behavior

Global identifier `chai` must be accessible.

## Actual Results

Only `chai.expect` and `chai.assert` are exported.

## Environment

* Operating System:  N/A
* Ethereum client: N/A
* Truffle version (`truffle version`): Truffle v5.0.21 (core: 5.0.21)
* node version (`node --version`): N/A
* npm version (`npm --version`):  N/A

**Comment by gnidan** [comment]
After some discussion, I think the better approach is for users to `require("chai")` manually. Putting it in the global space feels like scope creep. We'd like to remove globals from `truffle test`, ultimately.

Closing this, but feel free to offer a counter-argument. (I don't feel too strongly). Anyway, thanks for raising this!

**Comment by andvgal** [comment]
@gnidan  It seems @schuma7 and me could not emphasis on the problem: Truffle bundles own `chai` copy and it is NOT importable through `require('chai')`. So, it's not possible to install custom plugins.

```
Error: Cannot find module 'chai'
    at Function.Module._resolveFilename (module.js:548:15)
    at Function.Module._load (module.js:475:25)
    at Module.require (module.js:597:17)
    at require (internal/module.js:11:18) 
...

**Comment by CruzMolina** [comment]
Thanks for that info @andvgal . Reopening since it sounds like we should proceed with removing the injected `chai` globals.

**Comment by andvgal** [comment]
@CruzMolina as Truffle uses a bit own `contract()` API for test declaration, I doubt it's possible to just remove injected `chai`. Also, it would break a lot of projects in field.

As there is zero benefit of locally installed `chai` import in current tests, I believe there will be no conflict. So, injecting this global won't break the majority of projects. So, it can be a safe and quick solution.

**Comment by eggplantzzz** [comment]
@andvgal I could import it into my test files just fine. I just installed it with npm and added it to the dependencies section in the `package.json`. Afterwards I was able to require it in my js tests. Can you tell me exactly what you are doing when you get the `module not found` error?

**Comment by andvgal** [comment]
@eggplantzzz I am not installing `chai` separately as the one used by Truffle is not getting exposed. If you install a separate `chai` instance, you need to override `assert` and `except` globals to support custom plugins.

Of course, it is a solution, but it would just break all use cases in field. Based on the semver requirements, such change should not happen neither for patch nor for minor releases.

In short, I support your idea of getting rid of embedded chai, but development culture requires to postpone that to Truffle 6.x

**Comment by d10r** [comment]
I support that. Example for why:
Since truffle switched to the more promise based web3 1.0, it's often useful to add [chai-as-promised](https://www.chaijs.com/plugins/chai-as-promised/) to chai in order to more elegantly handle promise returning methods in tests.  
Setting that up however requires access to the global `chai` instance ([example](https://github.com/lab10-coop/artis-bridge-contracts/blob/ARTIS/test/setup.js)). As a consequence many truffle projects bundle their own chai. That's something I don't feel comfortable to do, because it's not clear to me if/which isses may or may not arise by shadowing the pre-injected chai instance like that.

Originally I wanted to propose including `chai-as-promised` in the pre-injected chai. However then I found this issue and now believe that a clearer solution would be to either give access to the global chai object or un-bundle it from truffle altogether.

**Comment by eggplantzzz** [comment]
If there is anything actionable on this subject for version 6, then we'll have to have a discussion and talk about specifications for the work.

**Comment by Lohann** [comment]
I'm done... I was trying to use [chai-bn](https://www.chaijs.com/plugins/bn-chai/) plugin to make my life easier, but realized that it is impossible using truffle. I'll do the same as OpenZeppelin, throw truffle into the garbage and use [openzeppelin-test-environment](https://github.com/OpenZeppelin/openzeppelin-test-environment) for unit testing.

**Comment by nonseodion** [comment]
Funny, but I stumbled here because of that same issue Lohan.

**Comment by hswopeams** [comment]
I also have the same problem. I can't use the chain-bn or bn-chain plugins without installing chai separately, but I don't know what affect installing it separately will have on the truffle framework since it includes a version of chai already.

**Comment by Lohann** [comment]
Update: [openzeppelin-test-environment](https://github.com/OpenZeppelin/openzeppelin-test-environment) has been deprecated, but Truffle is still a garbage, now I'm using [hardhat](https://hardhat.org/) + [ether.js](https://github.com/ethers-io/ethers.js/) + [waffle](https://getwaffle.io/): https://hardhat.org/guides/waffle-testing.html

**Comment by MartinSchere** [comment]
Please accept the pull request. A framework that relies on BigNumbers should be able to provide a testing suite supporting them.

**Comment by taylor8294** [comment]
Having had to deal with this problem myself, and the lack of movement on this for over two years, I can really understand Lohann's frustration on this. Seems a really obvious omission on Truffle's part.

**Comment by xluckydegen** [comment]
Is there ay news about this issue? I also would like to use chai-bn to compare BN in the test framework

**Comment by xluckydegen** [comment]
> @xluckydegen I think as of right now we're not going to do anything about this. Can you not install chai-bn separately and import it yourself for your use-case?

This is exactly what I wanted to do but didn't figure out a way how to access chai object to import it. But in the end I decided to use hardhat instead of truffle because there were no such issues.

**Comment by gnidan** [comment]
Despite [my 2019 comment](https://github.com/trufflesuite/truffle/issues/2090#issuecomment-501380632) suggesting that Truffle probably should not expose its own chai globals, I think that, so long as we do provide these assertions, we probably should also provide the normal extensions (like chai-bn).
